File: julia/difference-of-squares/src/DifferenceOfSquares.jl
module DifferenceOfSquares

export square_of_sum, sum_of_squares, difference

@doc raw"""
    square_of_sum(n::Integer)::Integer

Square the sum of the first `n` positive integers.

```math
\begin{align*}
  \left(\sum_{x=1}^n x\right)^2
  &= \left(\frac{n(n + 1)}{2}\right)^2 \\
  &= \frac{n^2 + 2n^3 + n^4}{4}
\end{align*}
```
"""
square_of_sum(n::Integer)::Integer = evalpoly(n, (0, 0, 1, 2, 1)) ÷ 4

@doc raw"""
    sum_of_squares(n::Integer)::Integer

Sum the squares of the first `n` positive integers.

```math
\begin{align*}
  \sum_{x=1}^n x^2
  &= \frac{n(n + 1)(2n + 1)}{6} \\
  &= \frac{n + 3n^2 + 2n^3}{6}
\end{align*}
```
"""
sum_of_squares(n::Integer)::Integer = evalpoly(n, (0, 1, 3, 2)) ÷ 6

@doc raw"""
    difference(n::Integer)::Integer

Subtract the sum of squares from square of the sum of the first `n` positive integers.

```math
\begin{align*}
  \left(\sum_{x=1}^n x\right)^2 - \sum_{x=1}^n x^2
  &= \left(\frac{n(n + 1)}{2}\right)^2 - \frac{n(n + 1)(2n + 1)}{6} \\
  &= \frac{n^2 + 2n^3 + n^4}{4} - \frac{n + 3n^2 + 2n^3}{6} \\
  &= \frac{-2n - 3n^2 + 2n^3 + 3n^4}{12}
\end{align*}
```
"""
difference(n::Integer)::Integer = evalpoly(n, (0, -2, -3, 2, 3)) ÷ 12

end # module DifferenceOfSquares

File: prolog/difference-of-squares/difference_of_squares.pl
:- module(difference_of_squares, [square_of_sum/2, sum_of_squares/2, difference/2]).  /*

\PrologDialect{swiprolog}

\Predicate square_of_sum/2(+N:int, -Result:int).

The square of the sum of the first $n$ numbers, i.e.,
\(
    \displaystyle
    \left(\sum_{k=1}^{n} k\right)^{2} = \left(\frac{n(n + 1)}{2}\right)^{2}
\).

\PL*/
%!  square_of_sum(+N:int, -Result:int) is semidet.
%
%   The square of the sum of the first =N= natural numbers.
square_of_sum(N, Result) :-
    Result is ((N * (N + 1)) / 2)^2.
/*PL

\Predicate sum_of_squares/2(+N:int, -Result:int).

The sum of the squares of the first $n$ natural numbers, i.e.,
\(
    \displaystyle
    \sum_{k=1}^{n} k^2 = \frac{n(n + 1)(2n + 1)}{6}
\).

\PL*/
%!  sum_of_squares(+N:int, -Result:int) is semidet.
%
%   The sum of the squares of the first =N= natural numbers.
sum_of_squares(N, Result) :-
    Result is (N * (N + 1) * (2 * N + 1)) / 6.
/*PL

\Predicate difference/2(+N:int, -Result).

The difference between the square of the sum of the first $n$ natural numbers
and the sum of the squares of the first $n$ natural numbers, i.e.,
\(
    \displaystyle
    \left(\sum_{k=1}^{n} k\right)^{2} - \sum_{k=1}^{n} k^{2}.
\)

\PL*/
%!  difference(+N:int, -Result:int) is semidet.
%
%   The difference between the square of the sum of the first =N= natural
%   numbers and the sum of the squares of the first =N= natural numbers.
%
%   @see square_of_sum/2, sum_of_squares/2.
difference(N, Result) :-
    square_of_sum(N, SquareOfSum),
    sum_of_squares(N, SumOfSquares),
    Result is SquareOfSum - SumOfSquares.
/*PL

\EndProlog*/
